Get Wheel Encoders working under ChibiOS

peterbarker opened the pull request:
Unify everything under AP_HAL::GPIO's attach_interrupt

rmackay9 commented:
I've tested these changes under ChibiOS and NuttX and it works.

peterbarker commented:
I've flight-tested this on a v2450 - the radio seems to work, which is nice.

peterbarker commented:
@rmackay9 RSSI should be fixed (Works For Me!) - is your review good to be dismissed?

rmackay9 commented:
@tridge or @OXINARF, could you have a look at this?  I can confirm that functionally it seems to work but I'm perhaps not qualified to say whether technically the approach is good.

tridge commented on libraries/AP_HAL_ChibiOS/GPIO.cpp:
```diff
@@ -115,19 +117,53 @@ AP_HAL::DigitalSource* GPIO::channel(uint16_t pin)
+    if (!g) {
+        return false;
+    }
+    g->fn = fn;
```
I think g->fn should not change unless the attach succeeds

tridge commented on libraries/AP_HAL_ChibiOS/GPIO.cpp:
```diff
@@ -147,25 +183,16 @@ bool GPIO::_attach_interrupt(ioline_t line, AP_HAL::Proc p, uint8_t mode)
+
     palDisableLineEvent(line);
     palEnableLineEvent(line, chmode);
-    palSetLineCallback(line, p?pal_interrupt_cb:nullptr, (void*)p);
```
we need to check if there is a conflict with another pad, by calling pal_lld_get_line_event() and checking if the cb field is set

tridge commented on libraries/AP_HAL_ChibiOS/GPIO.cpp:
```diff
@@ -202,3 +229,16 @@ void pal_interrupt_cb(void *arg)
+        return;
+    }
+
+    (g->fn)(g->pin_num, palReadLine(g->pal_line), now);
```
we need to check if g->fn is set

tridge commented on libraries/AP_RPM/RPM_Pin.cpp:
```diff
@@ -37,108 +37,36 @@ AP_RPM_Pin::AP_RPM_Pin(AP_RPM &_ap_rpm, uint8_t instance, AP_RPM::RPM_State &_st
-            stm32_gpiosetevent(last_gpio, false, false, false, nullptr);
-        }
+        // detach from last pin
+        hal.gpio->attach_interrupt(last_pin,
```
we need some indication of an error when this fails

tridge commented on libraries/AP_RSSI/AP_RSSI.cpp:
```diff
@@ -203,44 +203,74 @@ float AP_RSSI::read_channel_rssi()
+                                bool,
+                                uint32_t),
+            AP_HAL::GPIO::INTERRUPT_BOTH)) {
+        // failed to attach interrupt
```
need to report this somehow

tridge replied:
plus, it looks like this will keep trying, even if the pin doesn't change

tridge commented on libraries/AP_WheelEncoder/WheelEncoder_Quadrature.cpp:
```diff
@@ -15,122 +15,76 @@
+        if (!hal.gpio->attach_interrupt(pin,
+                                        fn,
+                                        AP_HAL::GPIO::INTERRUPT_BOTH)) {
+            hal.console->printf("WEnc: Failed to attach interrupt to pin=%u\n", pin);
```
can we use a STATUSTEXT here?

peterbarker commented:
On Sun, 19 Aug 2018, Andrew Tridgell wrote:

> I think g->fn should not change unless the attach succeeds

Fixed.

> >      palDisableLineEvent(line);
>      palEnableLineEvent(line, chmode);
> -    palSetLineCallback(line, p?pal_interrupt_cb:nullptr, (void*)p);
> 
> we need to check if there is a conflict with another pad, by calling
> pal_lld_get_line_event() and checking if the cb field is set

We do, several lines back.  However, there was an pre-existing race 
condition here which I've fixed by moving the modifications to within the 
osalSysLock


> we need to check if g->fn is set

Added.

> +        // detach from last pin
> +        hal.gpio->attach_interrupt(last_pin,
> 
> we need some indication of an error when this fails

Added.

> In libraries/AP_RSSI/AP_RSSI.cpp:

> +            AP_HAL::GPIO::INTERRUPT_BOTH)) {
> +        // failed to attach interrupt
> 
> need to report this somehow

Added.

> +        // failed to attach interrupt
> 
> plus, it looks like this will keep trying, even if the pin doesn't change

Moved the assignment so it will only try the attach once.  It does mean 
things won't start magically working when you free up the interrupt, but 
is consistent with other places.

> +        if (!hal.gpio->attach_interrupt(pin,
> +                                        fn,
> +                                        AP_HAL::GPIO::INTERRUPT_BOTH)) {
> +            hal.console->printf("WEnc: Failed to attach interrupt to pin=
> %u\n", pin);
> 
> can we use a STATUSTEXT here?

Yes, I've changed this to use gcs().send_text(...) as in all the other 
places.

OXINARF commented on libraries/AP_HAL/utility/functor.h:
```diff
@@ -60,12 +60,23 @@ class Functor
     }
+    inline bool operator!=(const Functor<RetType, Args...>& rhs)
+    {
+        return _obj != rhs._obj || _method != rhs._method;
```
Or `return !(this == rhs);`

OXINARF commented on libraries/AP_HAL/utility/functor.h:
```diff
@@ -60,12 +60,23 @@ class Functor
     }
+    // invalidate the functor; after this the bool operator on the
+    // functor will return false.
+    void invalidate()
```
Looks unused?

OXINARF commented on libraries/AP_HAL_Linux/GPIO_BBB.cpp:
```diff
@@ -113,7 +113,9 @@ AP_HAL::DigitalSource* GPIO_BBB::channel(uint16_t n) {
+                                irq_handler_fn_t fn,
+                                INTERRUPT_TRIGGER_TYPE mode)
 {
     return true;
```
We should change this to false.

OXINARF commented on libraries/AP_HAL_Linux/GPIO_RPI.cpp:
```diff
@@ -122,7 +122,9 @@ AP_HAL::DigitalSource* GPIO_RPI::channel(uint16_t n)
+                                irq_handler_fn_t fn,
+                                INTERRUPT_TRIGGER_TYPE mode)
 {
     return true;
```
Same.

OXINARF commented on libraries/AP_HAL_Empty/GPIO.cpp:
```diff
@@ -28,8 +28,9 @@ AP_HAL::DigitalSource* GPIO::channel(uint16_t n) {
+bool GPIO::attach_interrupt(uint8_t interrupt_num,
+                            irq_handler_fn_t fn,
+                            INTERRUPT_TRIGGER_TYPE mode) {
     return true;
```
Same.

OXINARF commented on libraries/AP_HAL_SITL/GPIO.cpp:
```diff
@@ -53,8 +53,10 @@ AP_HAL::DigitalSource* GPIO::channel(uint16_t n) {
+                            irq_handler_fn_t fn,
+                            INTERRUPT_TRIGGER_TYPE mode)
+{
     return true;
```
Same.

OXINARF commented on libraries/AP_HAL_VRBRAIN/GPIO.cpp:
```diff
@@ -173,7 +173,9 @@ AP_HAL::DigitalSource* VRBRAINGPIO::channel(uint16_t n) {
+                                   irq_handler_fn_t fn,
+                                   INTERRUPT_TRIGGER_TYPE mode)
 {
     return true;
```
What happened here? This should have the same as PX4 or we break VRBrain boards!

OXINARF commented on libraries/AP_WheelEncoder/WheelEncoder_Quadrature.h:
```diff
@@ -47,17 +48,17 @@ class AP_WheelEncoder_Quadrature : public AP_WheelEncoder_Backend
     static void update_phase_and_error_count(bool pin_a_now, bool pin_b_now, uint8_t &phase, int32_t &distance_count, uint32_t &total_count, uint32_t &error_count);
```
The two methods above don't need to be static anymore, right?

OXINARF commented on libraries/AP_WheelEncoder/WheelEncoder_Quadrature.cpp:
```diff
@@ -15,122 +15,82 @@
-        last_pin_b = get_pin_b();
+static uint32_t interrupt_count_a = 0;
+static uint32_t interrupt_count_b = 0;
+static uint32_t errors = 0;
```
What are these used for? I can't find it's use anywhere.

OXINARF commented on libraries/AP_RSSI/AP_RSSI.h:
```diff
@@ -76,16 +76,20 @@ class AP_RSSI
 
     // PWM input
     static struct PWMState {
```
No need to be static anymore, right?

OXINARF commented on libraries/AP_WheelEncoder/WheelEncoder_Quadrature.cpp:
```diff
@@ -15,122 +15,82 @@
-            stm32_gpiosetevent(irq_state[instance].last_gpio_a, true, true, false, _state.instance==0 ? irq_handler0_pina : irq_handler1_pina);
+    // install interrupt handler on rising or falling edge of gpio for pin a
+    if (new_pin != 255) {
+        if (!hal.gpio->attach_interrupt(pin,
```
Can we make it a single if? Applies to others places too.

OXINARF commented on libraries/AP_RSSI/AP_RSSI.cpp:
```diff
@@ -203,44 +205,81 @@ float AP_RSSI::read_channel_rssi()
+// read the PWM value from a pin
+float AP_RSSI::read_pwm_pin_rssi()
+{
+#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VRBRAIN || CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
```
Again, I don't see a good reason to put board checks here, it's using a HAL method.

OXINARF commented on libraries/AP_HAL_PX4/GPIO.cpp:
```diff
@@ -249,8 +251,152 @@ AP_HAL::DigitalSource* PX4GPIO::channel(uint16_t n) {
+        return false;
+    }
+
+    // attach a distict interrupt handler for each different GPIO
```
distict -> distinct

OXINARF commented on libraries/AP_RSSI/AP_RSSI.cpp:
```diff
@@ -203,44 +205,81 @@ float AP_RSSI::read_channel_rssi()
-            pwm_state.last_gpio = pwm_state.gpio;
+    // detach last one
+    if (pwm_state.last_rssi_analog_pin) {
+        if (pwm_state.last_rssi_analog_pin &&
```
This is repeated.

OXINARF commented on libraries/AP_HAL/GPIO.h:
```diff
@@ -37,9 +32,25 @@ class AP_HAL::GPIO {
+    FUNCTOR_TYPEDEF(irq_handler_fn_t, void, uint8_t, bool, uint32_t);
+    virtual bool    attach_interrupt(uint8_t pin,
+                                     irq_handler_fn_t fn,
+                                     INTERRUPT_TRIGGER_TYPE mode) = 0;
```
We could make this return false by default as well, since only two HALs implement it.

OXINARF commented on libraries/AP_HAL_ChibiOS/GPIO.cpp:
```diff
@@ -115,19 +117,56 @@ AP_HAL::DigitalSource* GPIO::channel(uint16_t pin)
+/*
    Attach an interrupt handler to ioline_t
  */
 bool GPIO::_attach_interrupt(ioline_t line, AP_HAL::Proc p, uint8_t mode)
```
This one line method seems a bit overkill to be used in a single place?

OXINARF commented on libraries/AP_HAL_ChibiOS/GPIO.cpp:
```diff
@@ -202,3 +232,18 @@ void pal_interrupt_cb(void *arg)
+        // what?
+        return;
+    }
+    if (g->fn == nullptr) {
```
We should just do `if (!g->fn)`, comparing to nullptr makes it use a temporary (null) functor (compiler might be smart here, or not).

OXINARF commented on libraries/AP_HAL_ChibiOS/GPIO.cpp:
```diff
@@ -202,3 +232,18 @@ void pal_interrupt_cb(void *arg)
 
+void pal_interrupt_cb_functor(void *arg)
+{
+    const uint32_t now = AP_HAL::micros();
```
Nitpicking: getting time before knowing we are really needing it.

rmackay9 commented:
Thanks for the reviews and rework, merged!

OXINARF commented:
VRBrain boards have been broken... as I warned in my review...

peterbarker commented:
On Tue, 21 Aug 2018, Francisco Ferreira wrote:

> VRBrain boards have been broken... as I warned in my review...

And I completely missed somehow - sorry.

I've stated to Randy that the first time someone complains they need the 
functionality I'll fix it (I'll need to borrow a board to test it, but I 
think it will pretty much be a matter of copying the PixHawk code).

I'm hoping we kill the NuttX code entirely before it's needed.

peterbarker commented:
On Mon, 20 Aug 2018, Francisco Ferreira wrote:

.... a great review, which I answered but never sent, apparently.  Oops.

>  *  override missing in some HALs

Fixed.

>  *  I would prefer a detach_interrupt method

Added.

>  *  WheelEnconder library has same initialization issue that you found in
>     RPM library

Added a patch to fix it.

> I hope some squash can be done before merging.

Lots of squashing done.

> In libraries/AP_RPM/RPM_Pin.h:
> 
> > @@ -19,6 +19,7 @@
>  #include "RPM_Backend.h"
>  #include <Filter/Filter.h>
>  #include <AP_Math/AP_Math.h>
> +#include <AP_HAL/GPIO.h>
> 
> Looks unneeded here?

Will change.  It's needed in the .cpp

> Also lines 19-22 of .cpp file should be removed now.

Will do.

> +    if (pin < 50) {
> 
> Isn't this worthless? The mapping already took place in a separate
> function and if it failed, this method already returned...

No - just an abundance of caution.  Trying very hard not to index into 
pin_handlers with something silly.  Yes, the mapping function would have 
to be faulty for this to happen - but the same is true of the second check 
below, too!

> In libraries/AP_HAL/GPIO.h:
> 
> >      /* Interrupt interface: */
> -    virtual bool    attach_interrupt(uint8_t interrupt_num, AP_HAL::Proc 
> p,
> -            uint8_t mode) = 0;
> +    //                                ret , pin    , state,timestamp
> +    FUNCTOR_TYPEDEF(irq_handler_fn_t, void, uint8_t, bool, uint32_t);
> 
> What is the state? True and false mean what?

I'll add some explanatory text.

> +void PX4GPIO::irq_handler(uint8_t handler)
> +{
> +    const uint32_t now = AP_HAL::micros();
> +
> +    AP_HAL::GPIO::irq_handler_fn_t fn = pin_handlers[handler];
> 
> Should it be a reference?

Made it a reference.

> +    if (handler == nullptr) {
> +        gcs().send_text(MAV_SEVERITY_WARNING, "No handler for @%u", gpio)
> ;
> 
> @? Also, this should be a debug message, it's useless for an user.

Makes the message little more interesting.  If this is ever emitted we 
want to know about it - making it a debug message may make that not 
happen.

> +    const uint8_t pin = handler_offset_to_pin[handler];
> 
> Any reason why handler_offset_to_pin is needed? Looks like handler + 50 is
> always the same (and the opposite is done in the attach_interrupt method).

Fair call.  Removed it.  Initially I was concerned this mapping might not 
always be the case.

> +    AP_HAL::GPIO::irq_handler_fn_t irq_handler_fn;
> 
> Why does this library keep a copy of the functor? Other libraries don't

Originally storage was going to be done in the caller.

> and it's not necessary since a copy is made into the HAL, the copy kept
> here is never used.

Well... we always assigned from it.

Removed this storage.

> 
> In libraries/AP_WheelEncoder/WheelEncoder_Quadrature.cpp:
> 
> >  #include <AP_BoardConfig/AP_BoardConfig.h>
> +#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VR
> BRAIN
>  #include <board_config.h>
> 
> What are these used for? It looks like these includes can go away. Also,
> any reason to keep this limited to these HALs (AP_RPM isn't)?

Removed.

Note that the HALs are rather random about their attach_interrupt methods.

> +    if (new_pin != 255) {
> 
> (uint8_t) -1 would be nicer. Applies to other places too.

Changed.

> +    void invalidate()
> 
> Looks unused?

Removed.

>  /* Interrupt interface: */
> -bool GPIO_BBB::attach_interrupt(uint8_t interrupt_num, AP_HAL::Proc p, ui
> nt8_t mode)
> +bool GPIO_BBB::attach_interrupt(uint8_t interrupt_num,
> +                                irq_handler_fn_t fn,
> +                                INTERRUPT_TRIGGER_TYPE mode)
>  {
>      return true;
> 
> We should change this to false.

I'd much rather remove all of these pointless overrides!  But as a 
different PR.

> -bool VRBRAINGPIO::attach_interrupt(uint8_t interrupt_num, AP_HAL::Proc p,
>  uint8_t mode)
> +bool VRBRAINGPIO::attach_interrupt(uint8_t interrupt_num,
> +                                   irq_handler_fn_t fn,
> +                                   INTERRUPT_TRIGGER_TYPE mode)
>  {
>      return true;
> 
> What happened here? This should have the same as PX4 or we break VRBrain
> boards!

It was like that when I got here!

I did consider fixing this - but considering the expected lifetime I think 
it would be wasted work.

> In libraries/AP_WheelEncoder/WheelEncoder_Quadrature.h:
> 
> > @@ -47,17 +48,17 @@ class AP_WheelEncoder_Quadrature : public AP_WheelEn
> coder_Backend
>      static void update_phase_and_error_count(bool pin_a_now, bool pin_b_n
> ow, uint8_t &phase, int32_t &distance_count, uint32_t &total_count, uint32
> _t &error_count);
> 
> The two methods above don't need to be static anymore, right?

Sure.  But the maping from a/b to phase probably still makes sense as a 
static, so I just changed the update function.

> In libraries/AP_WheelEncoder/WheelEncoder_Quadrature.cpp:
> 
> > 
> -    // check if pin b has changed and initialise gpio event callback
> -    if (last_pin_b != get_pin_b()) {
> -        last_pin_b = get_pin_b();
> +static uint32_t interrupt_count_a = 0;
> +static uint32_t interrupt_count_b = 0;
> +static uint32_t errors = 0;
> 
> What are these used for? I can't find it's use anywhere.

Ah.  Stray debug.  Removed.

>      // PWM input
>      static struct PWMState {
> 
> No need to be static anymore, right?

True.

> 
> __________________________________________________________________________
> 
> In libraries/AP_WheelEncoder/WheelEncoder_Quadrature.cpp:
> 
> > 
> -        // install interrupt handler on rising or falling edge of gpio fo
> r pin a
> -        irq_state[instance].last_gpio_a = get_gpio(last_pin_a);
> -        if (irq_state[instance].last_gpio_a != 0) {
> -            stm32_gpiosetevent(irq_state[instance].last_gpio_a, true, tru
> e, false, _state.instance==0 ? irq_handler0_pina : irq_handler1_pina);
> +    // install interrupt handler on rising or falling edge of gpio for pi
> n a
> +    if (new_pin != 255) {
> +        if (!hal.gpio->attach_interrupt(pin,
> 
> Can we make it a single if? Applies to others places too.

I prefer it this way here ATM.... did change in in other places.

> +// read the PWM value from a pin
> +float AP_RSSI::read_pwm_pin_rssi()
> +{
> +#if CONFIG_HAL_BOARD == HAL_BOARD_PX4 || CONFIG_HAL_BOARD == HAL_BOARD_VR
> BRAIN || CONFIG_HAL_BOARD == HAL_BOARD_CHIBIOS
> 
> Again, I don't see a good reason to put board checks here, it's using a
> HAL method.

I was concerned about the random values being returned in the HALs for 
attach-interrupt.

> distict -> distinct

Will fix.

> +    if (pwm_state.last_rssi_analog_pin) {
> +        if (pwm_state.last_rssi_analog_pin &&
> 
> This is repeated.

Will fix.

> +    FUNCTOR_TYPEDEF(irq_handler_fn_t, void, uint8_t, bool, uint32_t);
> +    virtual bool    attach_interrupt(uint8_t pin,
> +                                     irq_handler_fn_t fn,
> +                                     INTERRUPT_TRIGGER_TYPE mode) = 0;
> 
> We could make this return false by default as well, since only two HALs
> implement it.

Sounds good - as a different PR.  I don't know why it was done this way in 
the first place....

>  bool GPIO::_attach_interrupt(ioline_t line, AP_HAL::Proc p, uint8_t mode)
> 
> This one line method seems a bit overkill to be used in a single place?

Eh.  Doesn't hurt, really.

> We should just do if (!g->fn), comparing to nullptr makes it use a
> temporary (null) functor (compiler might be smart here, or not).

Will fix.

> +void pal_interrupt_cb_functor(void *arg)
> +{
> +    const uint32_t now = AP_HAL::micros();
> 
> Nitpicking: getting time before knowing we are really needing it.

This is actually done on purpose to give as tight a timing as possible on 
the interrupt.  This call might actually be subbed out for a lower-level 
call at some stage.

Thanks for the review!

OXINARF commented:
@peterbarker 

> It was like that when I got here!
>
>I did consider fixing this - but considering the expected lifetime I think it would be wasted work.

It wasn't like that, VRBrain was working with code in libraries and now that it's in HAL it isn't because it wasn't implemented in it. If it was wasted work for VRBrain it was even more for PX4 (most boards in PX4 are supported in ChibiOS, none of VRBrain is), but PX4 got implemented - not sure what was the issue with just copying it over.

> I've stated to Randy that the first time someone complains they need the functionality I'll fix it (I'll need to borrow a board to test it, but I think it will pretty much be a matter of copying the PixHawk code).
>
> I'm hoping we kill the NuttX code entirely before it's needed.

I completely disagree with that. We keep breaking VRBrain boards then people complain it's broken and we either blame VRBrain guys or turn to them to fix it - that's completely unfair; it's one thing unknowingly break it, it's another when we know it and don't care to fix it.

It would be great if all boards were ported to ChibiOS soon, but I doubt that's going to happen, so leaving VRBrain broken is a very bad idea.

> Will change.  It's needed in the .cpp

I don't think it was, it already included AP_HAL.h which includes all HAL headers. It's not bad in the .cpp though, it will just pass over it.

> I'll add some explanatory text.

Thanks, although the text still doesn't explain what true and false mean; it also mentions names that aren't in the function declaration (not sure if we can put them in the functor macro, worth a try).
